scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java

```diff
@@ -0,0 +1,326 @@
+        Mockito.when(dataFileRepository.findById(DATA_FILE_ID))
+                .thenReturn(Optional.of((DataFile) dataFile));
+        Mockito.when(fileSystemService.createPath(USERNAME, TYPE,
+                DATASET_ID_STRING)).thenReturn(path);
+        Mockito.when(fileSystemService.createPath(USERNAME, TYPE, DATASET_ID_STRING))
+                .thenReturn(path);
+        Mockito.when(fileSystemService.deleteDirectoryOrFile(file)).thenReturn(true);
+
+        Assertions.assertDoesNotThrow(() -> datasetService.deleteFileFromDataset(DATASET_ID,
+                DATA_FILE_ID));
+
+        Mockito.verify(datasetRepository).save((Dataset) Mockito
+                .argThat(new DatasetServiceTest.DatasetArgMatcher((dataset))));
+
+        Mockito.verify(dataFileRepository).deleteById(DATA_FILE_ID);
+        File fileWithName = new File(file + File.separator + DATAFILE_NAME);
+        Mockito.verify(fileSystemService).deleteDirectoryOrFile(fileWithName);
+    }
+
+    @Test
+    public void test_deleteFileFromDataset_no_dataset() {
+        Mockito.when(datasetRepository.findById(DATASET_ID))
+                .thenReturn(Optional.empty());
+        Assertions.assertThrows(DatasetNotFoundException.class,
+                () -> datasetService.deleteFileFromDataset(DATASET_ID, DATA_FILE_ID));
+    }
+
+    @Test
+    public void test_deleteFileFromDataset_no_datafile() {
+        Mockito.when(datasetRepository.findById(DATASET_ID))
+                .thenReturn(Optional.of((Dataset) dataset));
+        Mockito.when(dataFileRepository.findById(DATA_FILE_ID))
+                .thenReturn(Optional.empty());
+        Assertions.assertThrows(DataFileNotFoundException.class,
+                () -> datasetService.deleteFileFromDataset(DATASET_ID, DATA_FILE_ID));
+    }
+
+    @Test
+    public void test_deleteFileFromDataset_no_file_on_server() throws FileSystemStorageException {
+        dataset.addFile((DataFile) dataFile);
+        Mockito.when(datasetRepository.findById(DATASET_ID))
+                .thenReturn(Optional.of((Dataset) dataset));
+        Mockito.when(dataFileRepository.findById(DATA_FILE_ID))
+                .thenReturn(Optional.of((DataFile) dataFile));
+        Mockito.when(fileSystemService.createPath(USERNAME, TYPE, DATASET_ID_STRING))
+                .thenReturn(path);
+        Mockito.doThrow(new FileSystemStorageException("FileSystemStorageException"))
+                .when(fileSystemService).createPath(USERNAME, TYPE, DATASET_ID_STRING);
+
+        Assertions.assertThrows(DatasetStorageException.class,
+                () -> datasetService.deleteFileFromDataset(DATASET_ID, DATA_FILE_ID));
+
+        Mockito.verify(datasetRepository).save((Dataset) Mockito
+                .argThat(new DatasetServiceTest.DatasetArgMatcher((dataset))));
+        Mockito.verify(dataFileRepository).deleteById(DATA_FILE_ID);
+    }
+
+    static class DatasetArgMatcher extends ArgumentMatcher<IDataset> {
```

Comment:
It seems that the `ArgumentMatchers` require a new instance of the inner class to work, but I was under the impression that it's good practice to make inner classes like this static.

I searched and found the following:

https://stackoverflow.com/questions/16147405/whats-the-advantage-of-making-an-inner-class-as-static-with-java

I also get a warning from IntelliJ when removing static.

There's one `ArgumentMatcher` for `Datasets` and one `ArgumentMatcher` for `DataFiles` in their respective test classes.

Reply:
A static inner class can be used outside of the enclosing class without the need for an object of the enclosing class. Since no other class should be using the argument matcher outside of the test class, there is no need to make it static. If another class would use it, I would move it to its own file. 

Inner classes are good for logical grouping of classes, e.g. MapEntry only makes sense in the context of a Map, so it makes sense to have it as an inner class. Since the test is the only class using the matcher class, there is no need to make it available outside and making it static kind of sends the message that it can be used by other classes. But it would actually make it less straight forward to maintain if someone else would reuse it for their test cases in another class. Does this make sense? Maybe making it private will satisfy IntelliJ?